File: go/sieve/sieve.go
package sieve

func Sieve(limit int) (primes []int) {
	bools := make([]bool, limit+1)
	for i := 2; i <= limit; i++ {
		if !bools[i] {
			primes = append(primes, i)
			for j := i * 2; j <= limit; j = j + i {
				bools[j] = true
			}
		}
	}
	return primes
}

File: zig/sieve/sieve.zig
const std = @import("std");

pub fn primes(buffer: []u32, comptime limit: u32) []u32 {
    var sieve = std.StaticBitSet(limit + 1).initEmpty();
    sieve.set(0);
    sieve.set(1);
    var idx: u32 = 0;
    var n: u32 = 2;
    while (n <= limit) : (n += 1) {
        if (sieve.isSet(n)) continue;
        buffer[idx] = n;
        idx += 1;
        var j = n * n;
        while (j <= limit) : (j += n) {
            sieve.set(j);
        }
    }
    return buffer[0..idx];
}
